docs/design/apiserver_storage.md

```diff
@@ -0,0 +1,163 @@
+
+- easy to troubleshoot: Users should be able to identify easily if the storage layer is the current bottleneck of their cluster and why. We should expose enough metrics and write understandable docs to tell users how to do the troubleshooting.
+
+- easy to support multiple generic storage backends: The storage layer should have a clear separation between operations on Kubernetes objects and generic key-value pairs. To implement a new key-value or database backend, developers should not have to write duplicated code to handle Kubernetes logic. Ideally, they do not even need to know anything about Kubernetes at all.
+
+- backward compatible: Any changes should not introduce backward compatibility issues across at least last two minor leases from the end-user perspective. Using the improved storage system with existing etcd3 backend should just work.
+
+## Direction
+
+We could evolve the current storage layer towards the goal we listed above. It is doable, but it might involve significant interface and implement changes. Also it requires continuous changes of the consumers of the current storage API. During this process, we also need to ensure everything works perfectly. There is no alpha, or beta since we will be changing the single critical path. This is risky, time consuming, and hard to revert.
+
+The solution we proposed here is to start a storage layer from scratch. As Daniel Smith [mentioned](https://github.com/kubernetes/kubernetes/pull/28508#issuecomment-232508836)
+
+> Just to (probably unnecessarily) reiterate my position, I think we're going to need side-by-side implementations of both old and new storage styles, so we shouldn't end up locked into whatever we do now.
+
+We also believe we should start a side-by-side storage implementation to reach the goal we listed. That way, we won’t be affected by any existing issues, and it gives us the opportunities to rethink the design from the ground up, to achieve the goals more clearly. The only risk is that the effort will be completely wasted, if we cannot reach the goals.
+
+## Proposed Storage API
+
+```go
+
+type Storage interface {
+	// Wait waits until the storage has seen any version >= given version.
+	Wait(ctx context.Context, version uint64)
+
+	// Put puts an object with a key based on the given conditions.
+	// See docs on Conditions for more details.
+	Put(ctx context.Context, key string, obj runtime.Object, conditions *Conditions) (cur runtime.Object, err error)
+
+	// Delete a key and its object based on the given conditions.
+	// See docs on Conditions for more details.
+	Delete(ctx context.Context, key string, conditions *Conditions) (cur runtime.Object, err error)
+
+	// Get gets the most recent version of a key.
+	// A key that was modified by this interface will be reflected on read after write succeeded.
+	// If no object exists on the key, it will return not found error.
+	Get(ctx context.Context, key string) (cur runtime.Object, err error)
```

Comment:
I guess it's not terrible for people to call wait and then call get, I
don't even think we need a helper.

Getting an older version of an object would be handy for e.g. the garbage
collector, but we're going to have to solve that problem in another way
because we can't wait for this. :)

On Tue, Aug 2, 2016 at 10:53 AM, Xiang Li <notifications@github.com> wrote:

> In docs/design/apiserver_storage.md
> <https://github.com/kubernetes/kubernetes/pull/29888#discussion_r73204010>
> :
>
> > +type Storage interface {
> > +	// Wait waits until the storage has seen any version >= given version.
> > +	Wait(ctx context.Context, version uint64)
> > +
> > +	// Put puts an object with a key based on the given conditions.
> > +	// See docs on Conditions for more details.
> > +	Put(ctx context.Context, key string, obj runtime.Object, conditions *Conditions) (cur runtime.Object, err error)
> > +
> > +	// Delete a key and its object based on the given conditions.
> > +	// See docs on Conditions for more details.
> > +	Delete(ctx context.Context, key string, conditions *Conditions) (cur runtime.Object, err error)
> > +
> > +	// Get gets the most recent version of a key.
> > +	// A key that was modified by this interface will be reflected on read after write succeeded.
> > +	// If no object exists on the key, it will return not found error.
> > +	Get(ctx context.Context, key string) (cur runtime.Object, err error)
>
> @lavalamp <https://github.com/lavalamp> Is there a potential use case
> that we want to actually go back to history to get the past version of a
> key? Or we can add a helper function in storage pkg called WaitAndGet,
> which does wait and then get? I feel that can make the core API cleaner.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/kubernetes/kubernetes/pull/29888/files/ffd653da7d6f10f2cbae8c8c9ac9e739c96d1e47#r73204010>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAnglo2Hx4rP9ttZJWgqNk6qG2-L47Cjks5qb4QRgaJpZM4JaGTq>
> .
>